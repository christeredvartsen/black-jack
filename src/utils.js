const gameState = {
  BOTH_WINS: 'both win',
  BOTH_LOSE: 'both lose',
  P1_WINS: 'player 1 wins',
  P2_WINS: 'player 2 wins',
  P1_DRAWS: 'player 1 to draw',
  P2_DRAWS: 'player 2 to draw',
};

const getCardValue = ({ value }) => {
  let result;

  switch (value.toLowerCase()) {
    case 'j': // continue
    case 'q': // continue
    case 'k': result = 10; break;
    case 'a': result = 11; break;
    default: result = parseInt(value, 10);
  };

  return result;
};

const getCardsScore = (cards) => cards.reduce((score, card) => score + getCardValue(card), 0);

const getGameState = (p1, p2) => {
  if (p1 === 21 && p2 === 21) {
    return gameState.BOTH_WINS;
  } else if (p1 > 21 && p2 > 21) {
    return gameState.BOTH_LOSE;
  } else if (p2 > 21 || p1 === 21) {
    return gameState.P1_WINS;
  } else if (p1 > 21 || (p1 >= 17 && p2 > p1)) {
    return gameState.P2_WINS;
  } else if (p1 < 17) {
    return gameState.P1_DRAWS;
  } else {
    return gameState.P2_DRAWS;
  }
};

export {
  getCardValue,
  getCardsScore,
  gameState,
  getGameState,
};