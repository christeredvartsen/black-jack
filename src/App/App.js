import React, { Component } from 'react';
import axios from 'axios';
import Board from '../Board/Board';
import Error from '../Error/Error';
import { getCardsScore, gameState as s, getGameState } from '../utils';

const DECK_URL = 'https://nav-deckofcards.herokuapp.com/shuffle';

const initialState = {
  error: undefined,
  deck: [],
  player1: {
    name: 'Christer',
    score: 0,
    cards: [],
  },
  player2: {
    name: 'Magnus',
    score: 0,
    cards: [],
  },
  gameState: s.P1_DRAWS,
};

class App extends Component {
  requestSource = axios.CancelToken.source();
  state = Object.assign({}, initialState);

  drawCard = (currentPlayer) => {
    const [card, ...rest] = this.state.deck;
    const playersCards = [...this.state[currentPlayer].cards, card];
    const newScore = getCardsScore(playersCards);
    const updatedPlayer = Object.assign({}, this.state[currentPlayer], {
      cards: playersCards,
      score: newScore,
    });

    const otherPlayer = currentPlayer === 'player1' ? 'player2' : 'player1';
    const gameState = currentPlayer === 'player1'
      ? getGameState(updatedPlayer.score, this.state[otherPlayer].score)
      : getGameState(this.state[otherPlayer].score, updatedPlayer.score);

    this.setState({
      deck: rest,
      [currentPlayer]: updatedPlayer,
      gameState,
    });
  };

  reset = () => {
    this.setState(Object.assign({}, initialState));
    this.startGame();
  };

  startGame = () => {
    axios.get(DECK_URL, {
      timeout: 5000,
      cancelToken: this.requestSource.token
    }).then(({ data: deck }) => {
      const [c1, c2, c3, c4, ...rest] = deck;
      const player1Cards = [c1, c2];
      const player1Score = getCardsScore(player1Cards);
      const player2Cards = [c3, c4];
      const player2Score = getCardsScore(player2Cards);

      const player1 = Object.assign({}, this.state.player1, {
        cards: player1Cards,
        score: player1Score,
      });
      const player2 = Object.assign({}, this.state.player2, {
        cards: player2Cards,
        score: player2Score,
      });
      const gameState = getGameState(player1Score, player2Score);

      this.setState({
        deck: rest,
        player1,
        player2,
        gameState,
      });
    }).catch((error) => {
      this.setState({
        error: error.message
      });
    });
  };

  componentDidMount() {
    this.startGame();
  }

  componentWillUnmount() {
    this.requestSource && this.requestSource.cancel('Request cancelled');
  }

  render() {
    return typeof this.state.error !== 'undefined' ? (
      <Error message={this.state.error} reset={this.reset} />
    ) : this.state.deck.length > 0 ?
        <Board
          gameState={this.state.gameState}
          reset={this.reset}
          deck={this.state.deck}
          player1={this.state.player1}
          player2={this.state.player2}
          drawCard={this.drawCard}
        />
        : <p>Loading deck...</p>;
  }
}

export default App;
