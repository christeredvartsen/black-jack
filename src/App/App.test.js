import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import each from 'jest-each';
import waitUntil from 'async-wait-until';
import mockAxios from 'axios';
import App from './App';
import { gameState as s } from '../utils';

describe('<App />', () => {
  /*
  each([
    [
      // Player 1 wins
      [
        { suit: 'HEARTS', value: 'A' },
        { suit: 'HEARTS', value: 'K' },
        { suit: 'HEARTS', value: '2' },
        { suit: 'HEARTS', value: '3' },
      ],
      'foo',
    ],
    // Player 2 wins
    //[],
  
    // Both win
    //[],
  
    // Both lose
    //[],
  ]).test('can play a game', async (deck, resultMessage) => {
    mockAxios.get.mockImplementationOnce(() => Promise.resolve({
      data: deck,
    }));
    nock.disableNetConnect();
    nock(DECK_URL)
    .get('/shuffle')
    .reply(200, deck);
    
    const app = mount(<App />);
    //console.log(app.state('deck'));
    await waitUntil(() => app.state('deck').length !== 0);
    // c
    console.log(app.text());
    
    // expect(mockAxios.get).toHaveBeenCalledTimes(1);
    // expect(app.find('h2').text()).toBe('asd');
    // app.update();
    
    console.log(app.text());
  });
  */

  /*
  test('handles axios errors', () => {
    const wrapper = shallow(<App />);
    console.log(wrapper.debug());
  });
  */

  each([
    [
      'player1',
      [],
      0,
      [
        { value: 'A', suit: 'SPADES' }
      ],
      [],
      [
        { value: 'A', suit: 'SPADES' }
      ],
      11,
    ],
    [
      'player2',
      [
        { value: 'A', suit: 'SPADES' }
      ],
      11,
      [
        { value: 'K', suit: 'HEARTS' },
        { value: 'J', suit: 'CLUBS' },
      ],
      [
        { value: 'J', suit: 'CLUBS' },
      ],
      [
        { value: 'A', suit: 'SPADES' },
        { value: 'K', suit: 'HEARTS' },
      ],
      21,
    ],
  ]).test('players draws card', (player, currentPlayerCards, currentScore, deck, expectedRemainingDeck, expectedPlayerCards, expectedScore) => {
    const app = mount(<App />);
    app.setState({
      deck,
      [player]: {
        name: 'Christer',
        score: currentScore,
        cards: currentPlayerCards,
      }
    });
    app.instance().drawCard(player);
    expect(app.state('deck')).toEqual(expectedRemainingDeck);
    expect(app.state(player).score).toBe(expectedScore);
    expect(app.state(player).cards).toEqual(expectedPlayerCards);
  });

  test('displays error message', () => {
    const tree = mount(<App />);
    tree.setState({ error: 'some error' });
    expect(tree.find('.errorMessage').text()).toBe('some error');
  });

  test('matches snapshot', () => {
    const tree = shallow(<App />);
    expect(
      toJson(tree)
    ).toMatchSnapshot();
  });

  test('can reset game', () => {
    const app = mount(<App />);
    app.setState({ error: 'some error' });
    app.instance().reset();
    expect(app.state('error')).toBe(undefined);
  });

  each([
    [
      [
        { value: 'A', suit: 'SPADES' },
        { value: 'A', suit: 'HEARTS' },
        { value: 'A', suit: 'CLUBS' },
        { value: 'A', suit: 'DIAMONDS' },
        { value: '2', suit: 'CLUBS' },
        { value: '3', suit: 'CLUBS' },
      ],
      s.BOTH_LOSE
    ],
    [
      [
        { value: 'A', suit: 'SPADES' },
        { value: 'K', suit: 'HEARTS' },
        { value: 'A', suit: 'CLUBS' },
        { value: 'K', suit: 'DIAMONDS' },
        { value: '2', suit: 'CLUBS' },
        { value: '3', suit: 'CLUBS' },
      ],
      s.BOTH_WINS
    ],
    [
      [
        { value: 'A', suit: 'SPADES' },
        { value: 'K', suit: 'HEARTS' },
        { value: 'A', suit: 'CLUBS' },
        { value: 'A', suit: 'DIAMONDS' },
        { value: '2', suit: 'CLUBS' },
        { value: '3', suit: 'CLUBS' },
      ],
      s.P1_WINS
    ],
    [
      [
        { value: 'A', suit: 'SPADES' },
        { value: 'A', suit: 'HEARTS' },
        { value: 'A', suit: 'CLUBS' },
        { value: 'K', suit: 'DIAMONDS' },
        { value: '2', suit: 'CLUBS' },
        { value: '3', suit: 'CLUBS' },
      ],
      s.P2_WINS
    ],
  ]).test('can start game', async (deck, expectedGameState) => {
    mockAxios.get.mockImplementationOnce(() => Promise.resolve({
      data: deck
    }));
    const app = mount(<App />);
    await waitUntil(() => app.state('deck').length !== 0);
    expect(app.state('gameState')).toBe(expectedGameState);
  });
});