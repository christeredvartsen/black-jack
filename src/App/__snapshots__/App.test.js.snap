// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<p>
  Loading deck...
</p>
`;
